Some initial refactoring

archermarx opened the pull request:
Addresses #121  in part, as well as making some other changes to the internals. 
The main goal of this is to remove the vestiges of DifferentialEquations.jl. Previously, we had used that library for timestepping, and the entire design of the code was based on that. However, that led to some awkward design choices. Now that we no longer use DifferentialEquations, I would like to restructure the code in a way that makes more sense. This PR is a draft for now.

# What's changed
- The electron energy is no longer included in the state vector `U`. This saves about 3% of the computational time as we have reduced the number of rows we iterate through by 1
- The electron pressure coupled scheme has been removed. While nice, it saw very little use and was restricted to singly-charged ions only. Including it required a lot of special-casing that made the code longer and more difficult to follow.
- The test dependencies have been moved from test/Project.toml to the [extras] section of Project.toml. I have also pruned the test deps (removing `Plots` and others) so that the only heavy test dep is `Symbolics`, which we use for OVS.
- Some unused arguments have been removed from the stage limiter
- Remove the whole `ODEProblem` thing and associated structure
- Remove references to DifferentialEquations from docs

# Still to come (in later PRs)
- Restructure `params` and `cache`. These were designed to work with the DifferentialEquations.jl interface, but it's difficult to see what's in there. Ideally, we'd replace with more descriptively-named structures (perhaps "simulation" and "data"?). This would be a pretty large breaking change, so perhaps we could include methods for accessing things the "old way" for a while
- As part of the above, the Solution object could be made more useful with more descriptive fields. Having most of the important plasma properties hidden behind params.cache is not an ideal state of affairs
- Fix restarts (maybe hold off until structure is more stable)

archermarx commented:
I'm gonna mark this ready for review. @DecBrick, could you take a look?

DecBrick commented on docs/src/run.md:
```diff
@@ -136,13 +136,13 @@ tasks.
 ### The `Solution` object
 
-Running a simulation returns a `HallThruster.Solution` object. This mimics a DifferentialEquations `DiffEqSolution` object and has the same fields:
+Running a simulation returns a `HallThruster.Solution` object, which has the the same fields:
```
Might be missing context, but the "same fields" seems to still be vestigial.

archermarx replied:
Yep, it is. Fixed!

DecBrick commented on src/numerics/flux.jl:
```diff
@@ -304,13 +257,9 @@ function compute_fluxes!(F, UL, UR, U, params; apply_boundary_conditions = false
             # Sound speeds
-            aL = sqrt((charge_factor * TeL + γ * kB * TL) / mi)
-            aR = sqrt((charge_factor * TeR + γ * kB * TR) / mi)
+            aL = sqrt((γ * kB * TL) / mi)
```
As the ions currently only have a single temperature, is it worth leaving this in for if an ion temperature model is eventually implemented?

archermarx replied:
Yeah, that's what I'm currently thinking

DecBrick commented on src/wall_loss_models/constant_sheath_potential.jl:
```diff
@@ -14,7 +14,8 @@ function wall_power_loss(model::ConstantSheathPotential, U, params, i)
     ne = cache.ne[i]
-    ϵ = U[index.nϵ, i] / ne
+    nϵ = cache.nϵ[i]
+    ϵ = nϵ / ne
```
Seems like nϵ is only called once in this file, could save an allocation unless it's a setup for future changes

archermarx replied:
naming a variable doesn't cause a memory allocation. we still need to compute \epsilon in order to get the wall power loss here

DecBrick commented on test/runtests.jl:
```diff
@@ -80,3 +90,54 @@ end
         end
     end
 end
+
```
massive comment left at the bottom

archermarx replied:
removed, thanks

DecBrick reviewed (approved):
Looks good so far aside from the comments below. 

For future steps, with regards to the params structure, it may be good to have a method to be able to call all of its keys in a concise manner from a solution object to allow the user to be able to see what properties are saved. A native way to do this in Julia may exist,  but I haven't encountered it so far.

archermarx commented:
> Looks good so far aside from the comments below.
> 
> For future steps, with regards to the params structure, it may be good to have a method to be able to call all of its keys in a concise manner from a solution object to allow the user to be able to see what properties are saved. A native way to do this in Julia may exist, but I haven't encountered it so far.

i think `fieldnames(typeof(sol.savevals[1]))` would work, but i could add a `saved_properties` method that does that for you

DecBrick commented:
Well it does, maybe not do a full method but make a note about that in the documentation as the documentation on what fields the object has may not always be up to date.

archermarx commented:
I've added a `HallThruster.saved_fields()` function which will list all fields that get saved to `sol.savevals`. I've also updated the jupyter notebook to call this instead of using the static list of fields, and the docs so that this field is mentioned.
